### PR TITLE
Log training metrics to visualize them via tensorboard

### PR DESCRIPTION
**Pull request by tabergma**
**Proposed changes**:
Related to https://github.com/RasaHQ/research/issues/56

Add option ``tensorboard_log_directory`` to ``DIETClasifier``, ``ResponseSelector`` and
``TEDPolicy``.

By default ``tensorboard_log_directory`` is ``None``. If a valid directory is provided,
metrics are written during training. After the model is trained you can take a look
at the training metrics in tensorboard. Execute ``tensorboard --logdir <path-to-given-directory>``.

We also write down a model summary (layers, input size, type) in the provided directory.

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/utils/tensorflow/models.py [hunk 1]
```diff
@@ -109,12 +142,17 @@ def fit(
                         epoch_batch_size,
                         False,
                     )
-                    val_results = self._get_metric_results(prefix="val_")
+                    val_results = self._get_metric_results(
+                        epoch, self.test_summary_writer, prefix="val_"
```
Comment (dakshvar22) [inline rasa/utils/tensorflow/models.py]: We don't need the prefix, if we are creating the a separate summary writer for validation set. I think it's definitely better to create a separate summary writer for validation set as you can overlay train and val metrics then. so lets remove the `val_` prefix.

File: changelog/5422.feature.rst [hunk 2]
```diff
@@ -0,0 +1,7 @@
+Add option ``tensorboard_log_directory`` to ``DIETClasifier``, ``ResponseSelector`` and
+``TEDPolicy``.
+
+By default ``tensorboard_log_directory`` is ``None``. If a valid directory is provided,
```
[the remark this answers is not in the document]
Comment (tabergma) [reply changelog/5422.feature.rst]: Not sure if we need that. I guess, most users will not make use of this functionality as they are not interested in the training curves. Creating a directory by default might confuse them as they don't know what the data inside the directory represent.

File: changelog/5422.feature.rst [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+metrics are written during training. After the model is trained you can take a look
+at the training metrics in tensorboard. Execute ``tensorboard --logdir <path-to-given-directory>``.
+Make sure tensorboard is installed (``pip install tensorboard``).
```
[the remark this answers is not in the document]
Comment (tabergma) [reply changelog/5422.feature.rst]: Good point. Will check if it comes with tensorflow by default and remove it here.

File: rasa/utils/tensorflow/models.py [hunk 4]
```diff
@@ -35,6 +44,30 @@ def __init__(self, random_seed: Optional[int] = None, **kwargs) -> None:
...
+        self._set_up_tensorboard_writer(tensorboard_log_dir)
+
+    def _set_up_tensorboard_writer(
+        self, tensorboard_log_dir: Optional[Text] = None
+    ) -> None:
+        if tensorboard_log_dir is not None:
+            current_time = datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
+
+            train_log_dir = (
+                f"{tensorboard_log_dir}/logs/gradient_tape/{current_time}/train"
```
[the remark this answers is not in the document]
Comment (tabergma) [reply rasa/utils/tensorflow/models.py]: No. We can remove it.

**Comment by tabergma** [comment]
Just to be sure: How would you log on minibatch level? Would you have a counter that increases over time so that you see all steps in one plot? For example, if the first minibatch has 64 steps, the second minibatch would start with step 65 in the plot.

**Comment by dakshvar22** [comment]
@tabergma Just tried on the scaffold project dataset - The steps for test logger do not correspond to the steps for the train logger. For example, if `evaluate_after_every_epochs` is set to 20, then the steps for test logger should be 20,40,60 rather than 1,2,3. Sharing the step size would directly give an indication of which epoch does the validation loss correspond to. Otherwise the user has to calculate that on their own. Attaching a screenshot - 
Screenshot 2020-03-17 at 17 33 31
If this is easy enough to change, I would suggest we change it now.

**Comment by tabergma** [comment]
This only happens if you set `tensorboard_log_level` to `minibatch` correct? So in that case, training metrics should be plotted after every step, but evaluation metrics just after every epoch?

**Comment by dakshvar22** [comment]
Yes, setting it to `epochs` works fine. For your question regarding `minibatch` -  IMO they should be plotted after every `evaluate_after_every_epochs` steps. So if `evaluate_after_every_epochs` is set to 2, and number of minibatches in training set is 20 and number of minibatches in validation set is 2 then steps for test logger would be `40,41,80,81,120,121`

**Comment by tabergma** [comment]
Update: 
Do you mean something like this? 
Screenshot 2020-03-17 at 13 31 42

**Comment by dakshvar22** [comment]
Yes, there should be some inherent corresponding steps.

File: rasa/utils/tensorflow/models.py [hunk 5]
```diff
@@ -109,12 +161,21 @@ def fit(
                         epoch_batch_size,
                         False,
                     )
-                    val_results = self._get_metric_results(prefix="val_")
+                    step = epoch if self.tensorboard_log_on_epochs else training_steps
+                    val_results = self._get_metric_results(
```
Comment (dakshvar22) [inline rasa/utils/tensorflow/models.py]: Just tried it again, setting `tensorboard_log_level` to `minibatch` makes training logs at minibatch level but test logs at epoch level because of this function is called just once? Just confirming.

Comment (dakshvar22) [reply rasa/utils/tensorflow/models.py]: We should make it consistent actually, if it is set to `minibatch` then both train and test logging should be at minibatch level. I tested with this config - 
```
- name: DIETClassifier
    epochs: 75
    batch_size: 2
    tensorboard_log_directory: "./tb_logs"
    tensorboard_log_level: "minibatch"
    evaluate_on_number_of_examples: 20
```
Test logs just have one record after every `evaluate_after_every_epochs` which is not possible because there should `20/2` validation minibatches.

Comment (tabergma) [reply rasa/utils/tensorflow/models.py]: updated the code

File: rasa/utils/tensorflow/models.py [hunk 6]
```diff
@@ -35,6 +48,37 @@ def __init__(self, random_seed: Optional[int] = None, **kwargs) -> None:
...
+
+            current_time = datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
+
+            train_log_dir = f"{tensorboard_log_dir}/{current_time}/train"
+            test_log_dir = f"{tensorboard_log_dir}/{current_time}/test"
+
+            self.train_summary_writer = tf.summary.create_file_writer(train_log_dir)
+            self.test_summary_writer = tf.summary.create_file_writer(test_log_dir)
+
+            self.model_summary_file = (
```
Comment (dakshvar22) [inline rasa/utils/tensorflow/models.py]: Should we add the component name as a subfolder? Incase the same log directory is passed to all components, then the log directory becomes messy with just folders named after timestamps.
